filters/openpolicyagent/authorizewithregopolicy/authorizewithregopolicy.go

```diff
@@ -0,0 +1,163 @@
+package authorizewithregopolicy
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/zalando/skipper/filters"
+	"gopkg.in/yaml.v2"
+
+	"github.com/zalando/skipper/filters/openpolicyagent"
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/envoy"
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/util"
+)
+
+type Spec struct {
+	factory    openpolicyagent.OpenPolicyAgentFactory
+	configOpts []func(*openpolicyagent.OpenPolicyAgentInstanceConfig) error
+}
+
+func NewAuthorizeWithRegoPolicySpec(factory openpolicyagent.OpenPolicyAgentFactory, opts ...func(*openpolicyagent.OpenPolicyAgentInstanceConfig) error) Spec {
+	return Spec{
+		factory:    factory,
+		configOpts: opts,
+	}
+}
+
+func (s Spec) Name() string {
+	return "authorizeWithRegoPolicy"
+}
+
+const (
+	paramBundleName int = iota
+	paramEnvoyContextExtensions
+)
+
+func (s Spec) CreateFilter(config []interface{}) (filters.Filter, error) {
+	var err error
+
+	sargs, err := util.GetStrings(config)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(sargs) < 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	if len(sargs) > 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	bundleName := sargs[paramBundleName]
+
+	configOptions := s.configOpts
+
+	if len(sargs) > 1 {
+		envoyContextExtensions := map[string]string{}
+		err = yaml.Unmarshal([]byte(sargs[paramEnvoyContextExtensions]), &envoyContextExtensions)
+		if err != nil {
+			return nil, err
+		}
+		configOptions = append(configOptions, openpolicyagent.WithEnvoyContextExtensions(envoyContextExtensions))
+	}
+
+	opaConfig, err := openpolicyagent.NewOpenPolicyAgentConfig(configOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	opa, err := s.factory.NewOpenPolicyAgentEnvoyInstance(bundleName, *opaConfig, s.Name())
+
+	if err != nil {
+		return nil, err
+	}
+
+	return authorizeWithRegoPolicyFilter{
+		opa: opa,
+	}, nil
+}
+
+type authorizeWithRegoPolicyFilter struct {
+	opa *openpolicyagent.OpenPolicyAgentInstance
+}
+
+func (f authorizeWithRegoPolicyFilter) Request(fc filters.FilterContext) {
+	start := time.Now()
+
+	req := fc.Request()
+	span, ctx := f.opa.StartSpanFromContext(req.Context())
+	defer span.Finish()
+
+	authzreq := envoy.AdaptToEnvoyExtAuthRequest(req, f.opa.InstanceConfig().GetPolicyType(), f.opa.InstanceConfig().GetEnvoyContextExtensions())
+
+	result, err := f.opa.Eval(ctx, authzreq)
+
+	if err != nil {
+		f.opa.RejectInvalidDecisionError(fc, span, result, err)
+		return
+	}
+
+	f.opa.Logger().WithFields(map[string]interface{}{
+		"query":               f.opa.EnvoyPluginConfig().ParsedQuery.String(),
+		"dry-run":             f.opa.EnvoyPluginConfig().DryRun,
+		"decision":            result.Decision,
+		"err":                 err,
+		"txn":                 result.TxnID,
+		"metrics":             result.Metrics.All(),
+		"total_decision_time": time.Since(start),
```

Comment:
I think for a debug log that nobody enables a timer is a bit expensive. I would prefer to expose a metric with the same value, but on the other hand we also measure filter runtime in the proxy so I would question the time measurement here.

Reply:
Yes, I would also prefer to remove the debug logs. You can also still enable console decision logging via OPA so there is little benefit anyways. This currently just mimics what the ops Envoy plugin logs.

As for the metrics: Yes that is sth that I also had in mind. We can just measure the evaluation time which is a bit less than the proxy timing and handling the response around and at least gives insights into the latency introduced by policy evaluation itself. 

So I'd propose:
- Remove the debug logging
- Add a metric for eval times per filter instance (identified by "bundle name"), that way we can split the eval latencies per bundle as eval times between two bundles are not really comparable.

wdyt?

Reply:
Sounds like a great proposal. I guess the number of different "bundle name" will not be a huge number (something like 1-2 by application on average).

Reply:
Typically you'd have one "bundle" per application or even coarser with a normal discovery setup (the bundle name would identify the discovery bundle that then calculates the list of actual bundles). So my guess is that you can end up with 10s per Skipper deployment

Reply:
Seems to be good to go